FOCUS #104: ConsistencyHandling or StringFormat attribute (#396)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/string_handling.md

# String Handling

Columns that capture String values conforming to specified requirements foster data integrity, interoperability, and consistency, improve data analysis and reporting, and support more reliable data-driven decision-making.

> All columns capturing a String value, defined in the [FOCUS](#glossary:finops-cost-and-usage-specification) specification, MUST follow the requirements listed below. Custom String value capturing columns SHOULD adopt the same requirements over time.

Reviewer: ```suggestion
All columns capturing a String value, defined in the FOCUS specification, MUST follow the requirements listed below. Custom String value capturing columns SHOULD adopt the same requirements over time.
```

Author: As suggested, removed links in all attributes specs.

## Attribute ID

StringHandling

## Attribute Name

String Handling

## Description

Requirements for String-capturing columns appearing in billing data.

## Requirements

* String values MUST maintain the original casing, spacing, and other relevant consistency factors as specified by providers and end-users.

> * Changes to mutable string values (e.g., resource names) MUST be accurately reflected in charges incurred after the string value change and MUST NOT alter records incurred before the change, preserving data integrity and auditability for past billing periods.

Reviewer 2: I don't have a valid example at hand, but I'd imagine that this might be hard for providers to follow.  In the case that a mutable resource's name is changed and an adjustment is created to fix a charge for this resource from a month ago, I doubt the provider is still preserving the previous resource identification.  So, I'd change `MUST` to `SHOULD`, but I'm also not going to block on this comment.

Reviewer: I would rather keep as MUST and specify the condition where it's a SHOULD/Recommended etc. so we don't weaken the overall language

Author: Keeping the MUST. Edge case addressed as an exception. Hope it's ok.
@cnharris10 @marc-perreaut @udam-f2

Author: Addressed in https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/396#discussion_r1569631529 and approved at TF-1/TF-2 meeting

> * Changes to mutable string values (e.g., resource names) MUST be accurately reflected in charges related to subsequent costs incurred after the string value change and MUST NOT alter the original values in historical records, preserving data integrity and auditability for past billing periods.

Reviewer: ```suggestion
* Changes to mutable string values (e.g., resource names) MUST be accurately reflected in charges incurred after the string value change and MUST NOT alter records incurred before the change, preserving data integrity and auditability for past billing periods.
```

> * Immutable string values that refer to the same entity (e.g., resource identifiers, region identifiers, etc.) MUST remain consistent and unchanged across all billing periods.

Author: In response to the comment https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/396#issuecomment-2057950068

```suggestion
* Immutable string values that refer to the same entity (e.g., resource identifiers, region identifiers, etc.) MUST remain consistent and unchanged across all billing periods.
* Empty strings and strings consisting solely of spaces are not regarded as valid values for not-nullable string columns.
```

> * Immutable string values that refer to the same entity (e.g., resource identifiers) MUST remain consistent and unchanged across all billing periods.

Reviewer: ```suggestion
* Immutable string values that refer to the same entity (e.g., resource identifiers, region identifiers, etc.) MUST remain consistent and unchanged across all billing periods.
```

## Exceptions

> None

Author: In response to suggestion https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/396/files#r1567531933 by @cnharris10 

```suggestion
* When a record is provided after a change to a mutable string value and the [ChargeClass](#chargeclass) is "Correction", the record MAY contain the altered value.
```
